### PR TITLE
mds: trim cache regularly for standby-replay

### PR DESCRIPTION
**Pull request by batrick**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by batrick** [comment]
jenkins test api

File: src/mds/MDCache.cc [hunk 1]
```diff
@@ -13278,3 +13215,54 @@ void MDCache::handle_mdsmap(const MDSMap &mdsmap, const MDSMap &oldmap) {
...
+        return;
+      check_memory_usage();
+      if (mds->is_cache_trimmable()) {
+        dout(20) << "upkeep thread trimming cache; last trim " << since << " ago" << dendl;
+        bool active_with_clients = mds->is_active() || mds->is_clientreplay() || mds->is_stopping();
+        if (active_with_clients) {
+          trim_client_leases();
+        }
+        trim();
+        if (active_with_clients) {
```
Comment (sidharthanup) [inline src/mds/MDCache.cc:13240]: Isn't the if statement duplicated here?

Comment (batrick) [reply src/mds/MDCache.cc:13240]: We always want to run `trim` even if `!active_with_clients`. So, I don't think so?

**Comment by batrick** [comment]
rebased

File: src/mds/MDCache.cc (hunk 1 above)
Comment (sidharthanup) [reply src/mds/MDCache.cc:13240]: Got it! I missed that.

File: src/mds/MDSRank.h [hunk 2]
```diff
@@ -219,7 +219,7 @@ class MDSRank {
     bool allows_multimds_snaps() const { return mdsmap->allows_multimds_snaps(); }
 
     bool is_cache_trimmable() const {
-      return is_clientreplay() || is_active() || is_stopping();
+      return is_standby_replay() || is_clientreplay() || is_active() || is_stopping();
```
Comment (sidharthanup) [inline src/mds/MDSRank.h:222]: Was there any reason why standby_replay was left out of regular trimming before this?

Comment (batrick) [reply src/mds/MDSRank.h:222]: It was (and still is) initiated by another code path in MDLog. It just doesn't happen often enough to keep up with hot journals.

Comment (sidharthanup) [reply src/mds/MDSRank.h:222]: Ack.

**Review by sidharthanup — APPROVED** [review]
LGTM

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2021-04-07_02:12:41-fs-wip-pdonnell-testing-20210406.213012-distro-basic-smithi/

- https://tracker.ceph.com/issues/50215
    qa: "log [ERR] : error reading sessionmap 'mds2_sessionmap'"
- https://tracker.ceph.com/issues/49466
    qa: "Command failed on gibba030 with status 1: 'set -ex\nsudo dd of=/tmp/tmp.ZEeZBasJer'"
- https://tracker.ceph.com/issues/50216
    qa: "ls: cannot access 'lost+found': No such file or directory"
- https://tracker.ceph.com/issues/48773
    qa: scrub does not complete
- https://tracker.ceph.com/issues/49845
    qa: failed umount in test_volumes
- https://tracker.ceph.com/issues/50220
    qa: dbench workload timeout
- https://tracker.ceph.com/issues/50221
    qa: snaptest-git-ceph failure in git diff
- https://tracker.ceph.com/issues/50222
    osd: 5.2s0 deep-scrub : stat mismatch
- https://tracker.ceph.com/issues/50223
    qa: "client.4737 isn't responding to mclientcaps(revoke)"
- https://tracker.ceph.com/issues/50224
    qa: test_mirroring_init_failure_with_recovery failure

a bug in standby-replay is cluttering the results -- I'll rerun with the fix.

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2021-04-08_04:31:36-fs-wip-pdonnell-testing-20210408.024225-distro-basic-smithi/
https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20210408.142238 (with logic inversion / QA fix)

- https://tracker.ceph.com/issues/50246
    mds: failure replaying journal (EMetaBlob)
- https://tracker.ceph.com/issues/50250
    mds: "log [WRN] : Scrub error on inode 0x10000004506 (/client.0/tmp/clients/client3/~dmtmp/COREL) see mds.a log and `damage ls` output for details"

**Pull request merged**